### PR TITLE
Blazor Preview4 updates

### LINKED ISSUE #11837: [Preview 4] Razor components renaming
**For Preview 4 RTM**

https://github.com/aspnet/AspNetCore/issues/8931

Because sample and code example updates are concurrent with the naming changes, this should be worked the day that Preview 4 releases.

Details:

- [x] Rename the Razor Components template in .NET Core 3.0 back to “Blazor (server-side)”.
- [x] Rename the Blazor template to "Blazor (client-side)" for full clarity.
- [x] Rename `services.AddRazorComponents()` -> `services.AddServerSideBlazor()`
- [x] Rename `routes.MapComponentHub<TComponent>()` -> `routes.MapBlazorHub<TComponent>()`
- [x] Rename `components.server.js` back to `blazor.server.js` and `components.webassembly.js` back to `blazor.webassembly.js`.
- [x] Blazor on WebAssembly will no longer be experimental: Remove INCLUDE.
- [x] Get Started: One topic
  * Install SDK + templates
  * Tooling tabs: VS2019+VSIX; VSC+C# extension (**https://github.com/aspnet/AspNetCore.Docs/issues/11971#issuecomment-483620553**); check for other VSC
  * Template choice for desired experience
- [x] All *.razor* and all components in *Pages*.
- [x] Host & Deploy: Place the four topics under one node/folder.
- [x] Tutorial: Already a single topic: Clean it up for *.razor*, *Pages* folder, and tooling instructions.
- [ ] Common sample to P4 bits. **REGRESSION REPAIRS**
- [x] Remove the Razor Components (Blazor server-side) common sample.
- [x] "Razor components" app model language changed to "Blazor."
- [x] Recombine *Overviews* into one topic.
- [x] TOC work
- [x] Components in pages+views are interactive.
- [x] Blazor client-side get prerendering support for Preview 5. Update the language to reflect the push back to Pre5 for this scenario.
- [x] Update topic dates.
- [x] Mirror additional updates from the blog post in the docs.
- [x] *_Imports.razor*
- [ ] *Spit-'n-polish* ... it's the **spit** that really makes the docs shine. 🌞

Additional Info:

> We will still refer to the component model as Razor components (i.e. a .razor file will still be a Razor component), but when talking about the application model we are finding it very cumbersome to have different names.

> We're planning to keep the .razor extension for components. Components are based on Razor syntax, but components are handled differently by the Razor compiler, so they need a different file extension.

> What about the blazor hosted template? Answer: It stays as is. So there will be three Blazor templates: standalone client-side, ASP.NET Core hosted, and server-side

Convo on this issue resolves to ...

## Physical folder layout

* aspnetcore
  * blazor

`client-side` > `spa` > `blazor` 👋 *Bye!* 👋

## TOC

* Web apps
  * Blazor (after MVC)
  * Client-side development (after Blazor)

Client-side Development > SPA > Blazor 👋 *Bye!* 👋

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #11837 

**Here's the sample app :point_right:** https://github.com/guardrex/BlazorSamplePre4

* [Overview](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/index?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Get started](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/get-started?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Hosting models](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Build your first app](https://review.docs.microsoft.com/en-us/aspnet/core/tutorials/build-your-first-blazor-app?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Components](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/components?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Forms and validation](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/forms-and-validation?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Component libraries](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/class-libraries?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Layouts](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/layouts?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Dependency injection](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/dependency-injection?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Routing](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/routing?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [JavaScript interop](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/javascript-interop?view=aspnetcore-3.0&branch=pr-en-us-12048)
* [Debug](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/debug?view=aspnetcore-3.0&branch=pr-en-us-12048)
* Host and deploy
  * [Overview](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/blazor/index?view=aspnetcore-3.0&branch=pr-en-us-12048)
  * [Client-side](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/blazor/client-side?view=aspnetcore-3.0&branch=pr-en-us-12048)
  * [Server-side](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/blazor/server-side?view=aspnetcore-3.0&branch=pr-en-us-12048)
  * [Configure the Linker](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/blazor/configure-linker?view=aspnetcore-3.0&branch=pr-en-us-12048)

Common sample app runs, but ...

* I had to comment out the code-behind experience (`BlazorRocks.razor` and `BlazorRocksBase.cs`) to get the app to compile. The error (thrown dozens of times) is:

   `AZORGENERATE : error RZ3008: Tag helpers cannot target tag name '<global namespace>.BlazorRocksBase' because it contains a '<' character.`
* The TabSet example regressed: No styles now (but the code that toggles the hidden tab works).
* The Templated Components regressed: I had to comment out the `TableTemplate` and `ListViewTemplate` in `Pages/TemplatedComponents.razor`. The error thrown is ...

   `obj\Debug\netstandard2.0\Razor\Pages\TemplatedComponents.razor.g.cs(56,36): error CS0103: The name 'context' does not exist in the current context`

### CODE REVIEW

File: aspnetcore/blazor/get-started.md [hunk 1]
```diff
@@ -5,282 +5,119 @@ description: Learn how to get started with Blazor.
...
+      ```
+   
+   1. Open the *WebApplication1* folder.
+   1. When prompted by Visual Studio Code, add assets to build and debug the project:
+   
+      **Required assets to build and debug are missing from 'WebApplication1'. Add them?**
+   
+      Select **Yes**.
+   
+   1. Run the app using the Visual Studio Code debugger.
```
Comment (danroth27) [inline aspnetcore/blazor/get-started.md]: For client-side Blazor you can't run using the VSCode debugger

File: aspnetcore/blazor/index.md [hunk 2]
```diff
@@ -5,52 +5,39 @@ description: Explore ASP.NET Core Blazor, a way to build interactive client-side
...
-The connection used by Razor Components to communicate with the browser is also used to handle JavaScript interop calls.
-
-![Razor Components runs .NET code on the server and interacts with the Document Object Model on the client over a SignalR connection](index/_static/aspnet-core-razor-components.png)
-
-For more information, see <xref:blazor/hosting-models#server-side-hosting-model>.
-
 ## Components
 
-A *Razor Component* is a piece of UI, such as a page, dialog, or data entry form. Components handle user events and define flexible UI rendering logic. Components can be nested and reused.
+A *Razor component* in Blazor is an element of UI, such as a page, dialog, or data entry form. Components handle user events and define flexible UI rendering logic. Components can be nested and reused.
```
Comment (danroth27) [inline aspnetcore/blazor/index.md]:
```suggestion
A *component* in Blazor is an element of UI, such as a page, dialog, or data entry form. Components handle user events and define flexible UI rendering logic. Components can be nested and reused.
```

File: aspnetcore/blazor/index.md [hunk 3]
```diff
@@ -5,52 +5,39 @@ description: Explore ASP.NET Core Blazor, a way to build interactive client-side
...
 
-Components are .NET classes built into .NET assemblies that can be shared and distributed as NuGet packages. The class is normally written in the form of a Razor markup page with a *.razor* file extension (Razor Components) or Razor markup page with a *.cshtml* file extension (Blazor).
+Components are .NET classes built into .NET assemblies that can be shared and distributed as NuGet packages. The component class is usually written in the form of a Razor markup page with a *.razor* file extension.
```
Comment (danroth27) [inline aspnetcore/blazor/index.md]:
```suggestion
Components are .NET classes built into .NET assemblies that can be shared and distributed as NuGet packages. The component class is usually written in the form of a Razor markup page with a *.razor* file extension. Components in Blazor are sometimes referred to as Razor components.
```

**Comment by guardrex** [comment]
From the blog post, I still need to add ...

* Improved SignalR connection lifetime handling
* Improved disconnect & reconnect handling opportunities

Would you like me to continue added them to the *Server-side* section of the *Hosting models* topic where I added the other blog post sections?

**Comment by danroth27** [comment]
> Would you like me to add them to the Server-side section of the Hosting models topic where I added the other blog post sections?

That sounds reasonable. But I'd also like to get the Get Started and Overview updates in ASAP.

**Comment by guardrex** [comment]
Sure ... already done. I was collecting changes here as you go. I'll shoot those up now.

File: aspnetcore/blazor/common/samples/3.x/BlazorSample/Pages/BlazorRocksBase.cs [hunk 4]
```diff
@@ -1,7 +1,9 @@
+/* XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 using Microsoft.AspNetCore.Components;
 
 public class BlazorRocksBase : ComponentBase
```
Comment (danroth27) [inline aspnetcore/blazor/common/samples/3.x/BlazorSample/Pages/BlazorRocksBase.cs]: When I tried putting this type in a namespace (specifically BlazorSample) it worked.

**Comment by guardrex** [comment]
Looks like if I move `TableTemplate` and `ListViewTemplate` into the *Pages* folder, it fixes that problem.

**Comment by guardrex** [comment]
Now, I'm wondering if the TabSet example is broken for the same reason.

Do I need to do something special now for those components if they're over in the *Components* folder but being used by a component in the *Pages* folder?

**Comment by guardrex** [comment]
**_Confirmed!_** All problems fixed by moving the components in the *Components* folder over to the *Pages* folder. Should we just go with that?

**Comment by danroth27** [comment]
@guardrex Components have a namespace like normal .NET types, and in Preview4 we resolve component names using the in scope namespaces. The `BlazorSample.Components` namespace just needs to be in scope for you to use the components in that namespace. Just add `@using BlazorSample.Components` to your root *_Imports.razor* file.

**Comment by guardrex** [comment]
oh ... ok ... I'll fix it back up ... *properly* :smile:

File: aspnetcore/blazor/javascript-interop.md [hunk 5]
```diff
@@ -297,11 +297,11 @@ Console output in the browser's web developer tools:
 Hello, Blazor!
 ```
 
-## Share interop code in a Razor Component class library
+## Share interop code in a Razor component class library
```
Comment (danroth27) [inline aspnetcore/blazor/javascript-interop.md]: Replace "Razor Component class library" with "Blazor class library" in this section (`dotnet new blazorlib`). We don't have support for embedding static assets in Razor Class Libraries yet.

**Comment by guardrex** [comment]
Very good ... I'll merge it when the build completes.

I'll reread everything in the morning and see if I spot any 🙈 in the topics.

**Comment by guardrex** [comment]
No response ... he must be 🏃. **_Merging now!_**

**Pull request merged**